Added Salary form & breakdown view

TobiA34 opened the pull request:
- changed rootViewController to be SalaryFormViewController
- created salary form model
- created NI calculator to calculate national insurance
- created a pension contribution calculator to calculate pension contribution
- created a student loan calculator to calculate student loan
- created tax calculator to calculate tax
- created salary breakdown view controller
- created salary form view controller
- created salary form view model
- updated unit test
-  Added Validation Manager to validate text fields

https://github.com/TobiA34/Salary-App/assets/36420903/226d4e18-0ce8-4aff-a28d-a8b27f4a8dab

Validation:

<img width="430" alt="Screenshot 2023-07-12 at 16 39 26" src="https://github.com/TobiA34/Salary-App/assets/36420903/37689aa0-1d18-437b-b728-ad8ae53e4d31">
<img width="444" alt="Screenshot 2023-07-12 at 16 39 16" src="https://github.com/TobiA34/Salary-App/assets/36420903/cb100e4f-454e-4603-9062-a2f2ec529e67">
<img width="437" alt="Screenshot 2023-07-12 at 16 39 04" src="https://github.com/TobiA34/Salary-App/assets/36420903/e5a1da1d-0850-42e1-8c4b-ea593759230b">
<img width="445" alt="Screenshot 2023-07-12 at 16 38 53" src="https://github.com/TobiA34/Salary-App/assets/36420903/6cae5d76-efdf-492a-b3e4-8677408cac5d">

TobiA34 committed:
- Added Validation Manager to validate text fields
- changed rootViewController to be SalaryFormViewController
- created salary form model
- created NI calculator to calculate national insurance
- created pension contribution calculator to calculate pension contribution
- created student loan calculator to calculate student loan
- created tax calculator to calculate tax
- created salary break down view controller
- created salary form view controller
- created salary form view model
- updated unit test

tunds commented on Salary App/Features/SalaryForm/Model/Calculators/NICalculator.swift:
```diff
@@ -0,0 +1,22 @@
+
+        let ni = salary - 9568
+        let total = (ni * 0.12)/12
+        let doubleStr = String(format: "%.2f", total) // "3.14"
```
Why are you converting it into a string and then back into a double again? This is redundant since you just want to return the double anyways

tunds commented on Salary App/Features/SalaryForm/Model/Calculators/NICalculator.swift:
```diff
@@ -0,0 +1,22 @@
+       // £30,000 - £9,568 = £20,432
+      // (£20,432 * 0.12) / 12 = £204.32
+
+        let ni = salary - 9568
```
What is `9568`? Try to avoid magic numbers and create properties or leave comments explaining what these values represent so it's clear

tunds commented on Salary App/Features/SalaryForm/Model/Calculators/PensionContributionCalculator.swift:
```diff
@@ -0,0 +1,19 @@
+
+struct PensionContributionCalculator {
+    
+    func calculate(salary: Double, percentage: Double) -> Double? {
```
I think pensionPercentage would be a clearer parameter name here

tunds commented on Salary App/Features/SalaryForm/Model/Calculators/StudentLoanCalculator.swift:
```diff
@@ -0,0 +1,21 @@
+struct StudentLoanCalculator {
+    
+    func calculate(salary: Double) -> Double {
+        let rate = 0.09
```
There are two different student loan rates, here you may to research this cine it's not always 0.09. Take this into consideration and refactor this function

tunds commented on Salary App/Features/SalaryForm/Model/Calculators/StudentLoanCalculator.swift:
```diff
@@ -0,0 +1,21 @@
+    func calculate(salary: Double) -> Double {
+        let rate = 0.09
+        let studentLoan = (salary * rate) / 12
+        print(studentLoan)
```
Should this print be here?

tunds commented on Salary App/Features/SalaryForm/Model/Calculators/TaxCalculator.swift:
```diff
@@ -0,0 +1,33 @@
+import Foundation
+
+enum TaxBracket {
+    case PersonalAllowance
```
Enum cases should never start with a capital letter they should all be lower case i.e.

```

enum TaxBracket {
 case personalAllowance
....
}

Also are you even using this?

```

tunds commented on Salary App/Features/SalaryForm/View/SalaryBreakDownViewController.swift:
```diff
@@ -0,0 +1,261 @@
+    }()
+    
+    
+    private lazy var hourlyStackView: UIStackView = {
```
Rather than repeating this for hourly, weekly, monthly etc... Can you build a single component that you can reuse? You're repeating a lot of the same views here.

tunds commented on Salary App/Features/SalaryForm/View/SalaryBreakDownViewController.swift:
```diff
@@ -0,0 +1,261 @@
+    }()
+
+    
+    private lazy var niStackView: UIStackView = {
```
Can you turn these views i.e NI, Tax etc into a single reusable component. Since they're all similar to you can pass in some data and it will just layout/present it on the screen

tunds commented on Salary App/Features/SalaryForm/View/SalaryBreakDownViewController.swift:
```diff
@@ -0,0 +1,261 @@
+        niLbl.text = "National Insurance £\(salaryBreakdown.ni)"
+        studentLoanLbl.text = "Student Loan £\(salaryBreakdown.studentLoan)"
+        
+//        originalSalaryLbl.text = "Orignial salary: £\(originalSalary)"
```
Delete these comment if you don't need it

tunds commented on Salary App/Features/SalaryForm/View/SalaryBreakDownViewController.swift:
```diff
@@ -0,0 +1,261 @@
+        
+        originalSalaryLbl.text = "Orignial salary: £\(salaryBreakdown.originalSalary)"
+
+        salaryMonthlyLbl.text = "Monthly Total: £\(salaryBreakdown.monthlySalary)"
```
Rather than hard coding the `£` you have a number formatter for currency use this instead for all of these

tunds commented on Salary App/Features/SalaryForm/View/SalaryFormViewController.swift:
```diff
@@ -0,0 +1,336 @@
+    }()
+    
+    
+   private let validationManager = ValidationManager()
```
You're creating the same objects twice validationManager, taxCalculator, niCalculator, studentLoanCalulator. All of these also created inline.

If you want to init the salaryVm constant why don't you init everything in an init like

```


private let validationManager = ValidationManager()
   private let taxCalculator = TaxCalculator()
   private let niCalculator = NICalaculator()
   private let pensionCalculator = PensionContributionCalculator()

   private let salaryvm:  SalaryFormViewModel!

   init() {
   self.salaryVm = SalaryFormViewModel(taxCalculator: taxCalculator, niCalculator: niCalculator, studentLoanCalculator: studentLoanCalulator, pensionContributionCalculator: pensionCalculator)
    super.init()...
   
  
   }



```

tunds commented on Salary App/Features/SalaryForm/View/SalaryFormViewController.swift:
```diff
@@ -0,0 +1,336 @@
+        let salaryValue = Double(salaryTextField.text ?? "") ?? 0
+        let pensionValue = Double(pensionTextField.text ?? "") ?? 0
+         do {
+            try validationManager.isValid(texts: [salaryTextField.text ?? "" , pensionTextField.text ?? ""])
```
I wouln't be passing in an array of strings, this isn't the best approach... You should have one function called `isValid(input: FormInput)`. Then you do all of the validation in this one single function where you check each property and it throws an error all of the validation below needs be removed and instead you just do it in one function

tunds commented on Salary App/Features/SalaryForm/View/SalaryFormViewController.swift:
```diff
@@ -0,0 +1,336 @@
+            
+            navigationController?.pushViewController(sheetViewController, animated: true)
+            
+        } catch FormError.containsEmptyString {
```
Rather than having multiple catch you can just have one and use a switch statement.

```

catch { error in

switch error {
  case error is FormError.containsEmptyString:
     showAlert(title: "Error", message: error.localizedDescription, actionTitle: "Ok")
....
}


}

```

tunds commented on Salary App/Features/SalaryForm/View/SalaryFormViewController.swift:
```diff
@@ -0,0 +1,336 @@
+    }
+}
+
+extension UIViewController {
```
Move this into a seperate file, look into good naming conventions for extension files

tunds commented on Salary App/Features/ValidationManager/ValidationManager.swift:
```diff
@@ -0,0 +1,69 @@
+}
+
+extension FormError: LocalizedError {
+    var errorLocalDescription: String {
```
You shouldnt be creating your own computed property. When working with localizederror you can override the computed properties they offer check out this tutorial https://figure.ink/2021/07/18/practical-localized-error.html

tunds commented on Salary App/Features/ValidationManager/ValidationManager.swift:
```diff
@@ -0,0 +1,69 @@
+    case containsEmptyString
+    case formContainsLetters
+    case invalidSalary
+    case isValidPenContribution
```
This doesn't make sense how can a valid pension contribution be an error?

tunds replied:
Are these the only things that could go wrong?

tunds commented on Salary App/Features/ValidationManager/ValidationManager.swift:
```diff
@@ -0,0 +1,69 @@
+    }
+}
+
+struct ValidationManager {
```
This should only have one function that accepts a class where you validate all its properties and then throw an error when a condition isn't met.

tunds commented on Salary App/Features/ValidationManager/ValidationManager.swift:
```diff
@@ -0,0 +1,69 @@
+
+import Foundation
+
+enum FormError: Error {
```
You should have a computed property for the error title that you want to show in the alert

tunds commented on Salary App/Features/ValidationManager/ValidationManager.swift:
```diff
@@ -0,0 +1,69 @@
+
+     }
+    
+    func inValidNumber(values: [String]) throws {
```
Remove this function and do the validation in this function in the isValid function above based on my comments above.

tunds commented on Salary App/Features/ValidationManager/ValidationManager.swift:
```diff
@@ -0,0 +1,69 @@
+
+    }
+    
+    func isValidSalary(number: Double) throws{
```
Remove this function and do the validation in this function in the isValid function above based on my comments above.

tunds commented on Salary App/Features/ValidationManager/ValidationManager.swift:
```diff
@@ -0,0 +1,69 @@
+        }
+    }
+    
+    func isValidPenContribution(number: Double) throws {
```
Remove this function and do the validation in this function in the isValid function above based on my comments above.

tunds commented on SalaryAppTests/SalaryAppTests.swift:
```diff
@@ -9,42 +9,110 @@ import XCTest
+    func testWeelySalary() throws {
+    salary = 30000.0
+    let pensionContributions = 20.0
+    let salaryFormVM = SalaryFormViewModel(taxCalculator: TaxCalculator(), niCalculator: NICalaculator(), studentLoanCalculator: StudentLoanCalculator(), pensionContributionCalculator: PensionContributionCalculator())
```
The salary form vm should be created in setup and destroyed in teardown do this throughout the whole project

tunds commented:
Can you document all the functions please

TobiA34 committed:
- Added extension file for alert
- Added Horizontal stack view reusable component
- Updated NI calculator calculate function
- Updated pension contribution calculator function
- Added enum student loan plan and updated calculate function with switch and cases
- Documented TaxCalculator function
- Added SalaryHorizontalStackView on the monthly, yearly, daily, weekly, hourly, ni and orginal stack view in SalaryBreakDownViewController
- Added number formatter and updated all the UILabels to use number formatter  in SalaryBreakDownViewController
- Added array studentLoanPlans and student loan label, student loan stack view and student loan textfield in SalaryFormViewController
- Injected validation manager and calculator objects in SalaryFormViewController
- Updated constant monthly salary, daily salary, weekly salary and hourly salary in SalaryFormViewController
- Updated catch block to include switch statement which shows alert based on enum case in SalaryFormViewController
- Documented functions in salary form view model
- Updated  enum FormError to include error title and moved validation in isValid method
- Added salary form view model and student loan plan in the Salary App Tests
